File: Midtermlab3/MP13 – Detect Rows with Missing Values/MP13_missingValues.js
// MP13 - Detect rows with missing values

const fs = require("fs");
const readline = require("readline");

// Ask user for file path
const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout
});

rl.question("Enter CSV file path: ", function(path) {

  try {
    const data = fs.readFileSync(path, "utf8");
    const rows = data.split("\n");

    console.log("\nRows with missing values:");

    rows.forEach((row, index) => {
      const values = row.split(",");

      const hasMissing = values.some(v => v.trim() === "");

      if (hasMissing) {
        console.log(`Row ${index + 1}: ${row}`);
      }
    });

  } catch (error) {
    console.log("Error reading file:", error.message);
  }

  rl.close();
});

File: Midtermlab3/MP13 – Detect Rows with Missing Values/MP13_MissingValues.java
// MP13 - Detect rows with missing values in a CSV file

import java.io.*;
import java.util.*;

public class MP13_MissingValues {

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);

        System.out.print("Enter CSV file path (or press Enter for Sample_Data-Prog-2-csv.csv): ");
        String filePath = scanner.nextLine().trim();

        if (filePath.isEmpty()) {
            filePath = "Sample_Data-Prog-2-csv.csv";
            System.out.println("Using default file path: " + filePath);
        }

        File csvFile = new File(filePath);
        if (!csvFile.exists() || !csvFile.isFile()) {
            System.out.println("Error: File does not exist or is not a file: " + filePath);
            scanner.close();
            return;
        }

        int rowNumber = 0;
        boolean foundMissing = false;

        try (BufferedReader br = new BufferedReader(new FileReader(csvFile))) {
            System.out.println("\nRows with missing values:");

            String line;
            while ((line = br.readLine()) != null) {
                rowNumber++;
                if (line.trim().isEmpty()) {
                    continue; // skip empty lines
                }

                String[] values = line.split(",", -1); // preserve empty fields
                boolean missing = false;

                for (String value : values) {
                    if (value.trim().isEmpty()) {
                        missing = true;
                        break;
                    }
                }

                if (missing) {
                    foundMissing = true;
                    System.out.println("Row " + rowNumber + ": " + line);
                }
            }

            if (!foundMissing) {
                System.out.println("No rows with missing values found.");
            }

        } catch (IOException e) {
            System.out.println("Error reading file: " + e.getMessage());
        }

        System.out.println("\nProgram Finished.");
        scanner.close();
    }
}